logging: wake up log process thread only when log message happen

wentongwu opened the pull request:
Periodically wake up log process thread consume more power even no logging message in message list if system already in sleep or deep sleep state. With the help of added logging timer and semaphore, log process thread is woken up only when there is logging message.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

nordic-krch commented on subsys/logging/log_core.c:
```diff
@@ -1139,6 +1146,13 @@ void log_hexdump_from_user(struct log_msg_ids src_level, const char *metadata,
 
+static void log_process_thread_timer_expiry_fn(struct k_timer *timer)
+{
+	k_timer_stop(timer);
```
do you need stop here?

wentongwu replied:
solved, thanks

nordic-krch reviewed (commented):
I like this approach. I wonder if in that case we shouldn't change default for LOG_PROCESS_THREAD_SLEEP_MS to something lower (e.g. 100ms) to make logger more responsive. Also if sleep is lower then threshold could be increased imo.

wentongwu commented:
> I like this approach. I wonder if in that case we shouldn't change default for LOG_PROCESS_THREAD_SLEEP_MS to something lower (e.g. 100ms) to make logger more responsive. 

@nordic-krch sorry for the late response and thanks for the review. This patch doesn't change the wake up frequency of log process thread if always there is logging message. If want to more responsive, we should change threshold to lower value.

> Also if sleep is lower then threshold could be increased imo.

Do you mean dynamically change the threshold(LOG_PROCESS_THREAD_SLEEP_MS) accordingly?

wentongwu committed:
logging: wake up log process thread only when log message happen

Periodically wake up log process thread consume more power if system
already in sleep or deep sleep state. With the help of added logging
timer and semaphore, log process thread is woken up only when there
is logging message.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

tyrantlizard77 reviewed (commented):
@nordic-krch I have a question: why don't we get rid of CONFIG_LOG_PROCESS_THREAD_SLEEP_MS entirely, and just have the log process thread simply sleep on the semaphore, which will get given any time there is a new log message to process? did we try this at some point and find a reason not to do it?

In other words, why do we even need log_process_thread_timer?

nordic-krch commented:
@andrewboie 
>  just have the log process thread simply sleep on the semaphore

Performance was the reason to not do that. Logger in the best case on nrf devices takes ~4us to log a typical message. Adding semaphore call would double this time and one of the goals of the logger was to be as non-intrusive as possible.
